### DISCUSSION TITLE
Custom dialog boxes

### DISCUSSION THREAD

**Discussion by brandonkelly** [craftcms/cms#8402] [Ideas]
Native browser dialog boxes (alerts, confirms, etc.) feel pretty out of place in the Craft CP. Worse, some browsers give users the ability to disable future "similar" alerts, which can completely break CP functionality, like deleting things.

We should move to using custom dialogs in the CP - something like http://t4t5.github.io/sweetalert/

**Comment by brandonkelly** [comment]
Another benefit: It would enable us to have similar smart selection logic for the “Rename File” asset action that the Filename setting in asset editors have (where everything but the extension gets selected when you first click on it, etc.)

**Comment by narration-sd** [comment]
Yet another plus for this, which am sure you haven't missed -- in today's security environment, any generic-looking alert or dialog box immediately brings suspicion of malware. 

Plus anyway, the Chrome defaults are particularly ugly, and this is another area that can benefit from Brandon's js-ui prowess...

**Comment by andregoldstein** [comment]
I personally find sweetalert pretty over the top in design terms (doesn't fit too well with Craft's professional, understated design for me).

But the idea in general is a good one.

**Comment by RitterKnightCreative** [comment]
+1 for this. Safari/WebKit has standardized on iOS-style alerts as well, even on desktop,  and they look rather... plain.

Could be something Craft would bake in via Garnish? The UI dialogs in Craft themselves can't that much different... you even had the esc/return keys doing the right things too.  FWIW at least SweetAlert gets this right.

**Comment by proimage** [comment]
This has bugged me for a while now. Craft's UI is so refined (a co-worker recently described coming back to Craft after working in WP as being like going "from a clunker to a Porsche"), it's a shame that these native UI dialog boxes are still around, marring the experience. :)

**Comment by narration-sd** [comment]
Here's an idea that's worked well in some projects here, and could be extended for Craft.

When we need an alert, or a dialog, instead of popping a modal, native or not, we just enable a div to be exposed on the UI itself.

The effect is that the rest of the screen shifts down, so that the error or whatever now shows as part of, and in the style of the overall app UX.

It's great for alerts, loggings, etc., and would work for non-critical-to-event-flow dialogs, and it keeps everything looking in-the-app, and nice.

Modals...not so easy, but then Garnish has a fine ability for modals, certainly experienced. 

I'm thinking a moment if there's any interesting way to combine the two approaches, but actually probably not, nor needed. Probably this thought is triggered by something done as all of this in Vue, where the (true) modal content became a div always hidden on the main template, until teleported into the modal. But if I remember, using Garnish modals gets you something much like that anyway in the html of the page.

If the thought helps...

**Comment by proimage** [comment]
With Craft's recent focus on accessibility, I would hope and presume that any modal box would not be a meaningless `<div>`, but rather a [semantically-significant HTML5 `<dialog>` element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dialog).

**Comment by brandonkelly** [comment]
Yeah we need to fix the accessibility of our current JS modals/dialogs before we start adding new ones :)

**Comment by RitterKnightCreative** [comment]
FWIW I noticed that [sweet alert 2](https://sweetalert2.github.io/) is out and claims to have accessibility support.
